azurelinuxagent/common/cgroups.py

```diff
@@ -0,0 +1,698 @@
+        Create all required CGroups to track all metrics for an extension and its associated services.
+
+        :param str name: Full name of the extension to be tracked
+        :param CGroups cgroup: CGroup for the extension itself. This method will create it if none is supplied.
+        """
+        if not CGroupsTelemetry.is_tracked(name):
+            cgroup = CGroups.for_extension(name) if cgroup is None else cgroup
+            logger.info("Now tracking cgroup {0}".format(name))
+            CGroupsTelemetry.track_cgroup(cgroup)
+        if CGroups.is_systemd_manager():
+            if name in related_services:
+                for service_name in related_services[name]:
+                    CGroupsTelemetry.track_systemd_service(service_name)
+
+    @staticmethod
+    def is_tracked(name):
+        return name in CGroupsTelemetry._tracked
+
+    @staticmethod
+    def stop_tracking(name):
+        """
+        Stop tracking telemetry for the CGroups associated with an extension. If any system services are being
+        tracked, those will continue to be tracked; multiple extensions might rely upon the same service.
+
+        :param str name: Extension to be dropped from tracking
+        """
+        if CGroupsTelemetry.is_tracked(name):
+            del (CGroupsTelemetry._tracked[name])
+
+    @staticmethod
+    def set_delete_empty_groups(setting):
+        CGroupsTelemetry.delete_empty_groups = setting
+
+    @staticmethod
+    def collect_all_tracked():
+        """
+        Return a dictionary mapping from the name of a tracked cgroup to the list of collected metrics for that cgroup.
+
+        :returns: Dictionary of collected metrics, by cgroup
+        :rtype: dict(str: [(str, str, float)])
+        """
+        results = {}
+        for cgroup_name, collector in CGroupsTelemetry._tracked.items():
+            cgroup_name = cgroup_name if cgroup_name else "Agent+Extensions"
+            results[cgroup_name] = collector.collect()
+        return results
+
+    @staticmethod
+    def update_tracked(ext_handlers):
+        """
+        Track CGroups for all enabled extensions.
+        Track CGroups for services created by enabled extensions.
+        Stop tracking CGroups for not-enabled extensions.
+
+        :param List(ExtHandler) ext_handlers:
+        """
+        not_enabled_extensions = set()
+        for extension in ext_handlers:
+            if extension.properties.state == u"enabled":
+                CGroupsTelemetry.track_extension(extension.name)
+            else:
+                not_enabled_extensions.add(extension.name)
+
+        names = []
+        for name in CGroupsTelemetry._tracked.keys():
+            if CGroupsTelemetry.delete_empty_groups and name in not_enabled_extensions:
+                CGroupsTelemetry.stop_tracking(name)
+            else:
+                names.append("[{0}]".format(name))
+
+        now_tracking = " ".join(names)
+        if now_tracking != CGroupsTelemetry.was_tracked:
```

Comment:
I don't like this comparison.  You **appear** to be relying on a string to determine if things have changed or not.  I don't think that is a robust comparison.  Why isn't this a set compare instead?

Reply:
OK.

Reply:
Significantly refactored, switched to using a set.